### PR TITLE
Improve uri handling

### LINKED ISSUE #48: Metadata URL with values in Query String
Hi.

The url to get the metadata from my OData API requires setting the Api version in a parameter sent in the url query string.

http://[host]:[port]/odata/$metadata **?api-version=1.0**

In this case, when you try to configure the connected service, an error occurs because the extension tries to concatenate "$metadata" to the end of the URL.

How can I use this feature with the exposed scenario?

Thank you

Victor Perez

image

### PR DESCRIPTION
**Pull request by Nthemba**
This PR fixes the issue [Metadata URL with values in Query String · Issue #48 · OData/ODataConnectedService (github.com)](https://github.com/OData/ODataConnectedService/issues/48)
and is an add on to a contributor's (https://github.com/dr-consit) PR [Fix #48 by improving handling of Uris with/without "$metadata" suffix. by dr-consit · Pull Request #237 · OData/ODataConnectedService (github.com)](https://github.com/OData/ODataConnectedService/pull/237)

It ensures the OData Endpoint URI does not get distorted when the URI includes query and fragment segments. 
### Before fix 
A URI like 'http://localhost:5000/odata/$metadata?api-version=1.0' would have $metadata appended to the end of the URI which would make the OData service inaccessible.
### After fix
$metadata, query and fragment segments maintain the correct position and order.

### CODE REVIEW

**Comment by ElizabethOkerio** [comment]
@Nthemba resolve conflicts.

File: azure-pipelines.yml [hunk 1]
```diff
@@ -101,10 +101,10 @@ stages:
         publishLocation: 'Container'
 
     - task: PublishBuildArtifacts@1
-      displayName: 'Publish Artifact - VSIX2022plus'
```
Comment (habbes) [inline azure-pipelines.yml:104]: Why are these changes necessary?

File: src/Microsoft.OData.Cli/GenerateCommand.cs [hunk 2]
```diff
@@ -185,16 +186,15 @@ private async Task<int> HandleGenerateCommand(GenerateOptions options, IConsole
             catch (Exception ex)
             {
                 console.Error.Write(ex.Message);
-                return 0;
+                return 1;
             }
 
-            return 1;
```
Comment (habbes) [inline src/Microsoft.OData.Cli/GenerateCommand.cs:191]: I thought this was already changed in a different PR?

File: src/Microsoft.OData.CodeGen/Common/MetadataReader.cs [hunk 3]
```diff
@@ -24,36 +26,47 @@ public static class MetadataReader
...
             Stream metadataStream;
-            Uri metadataUri = new Uri(serviceConfiguration.Endpoint);
+            var metadataUri = new Uri(serviceConfiguration.Endpoint);
 
             if (!metadataUri.IsFile)
             {
                 var webRequest = (HttpWebRequest)WebRequest.Create(metadataUri);
+                if (serviceConfiguration.CustomHttpHeaders != null)
+                {
+                    var headerElements = serviceConfiguration.CustomHttpHeaders.Split(new[] { Environment.NewLine }, StringSplitOptions.RemoveEmptyEntries);
```
Comment (habbes) [inline src/Microsoft.OData.CodeGen/Common/MetadataReader.cs:58]: Not sure I understand how this relates to the issue the PR is fixing.

File: src/Microsoft.OData.CodeGen/Common/MetadataReader.cs [hunk 4]
```diff
@@ -79,32 +92,76 @@ public static Version GetMetadataVersion(ServiceConfiguration serviceConfigurati
...
+        public static Uri CleanMetadataUri(this Uri uri)
+        {
+            if (uri.Scheme == "http" || uri.Scheme == "https")
+            {
+                UriBuilder uriBuilder;
+
+                /// Evaluates to true if Query and Fragment properties are present in the Uri 
+                bool preserveQueryAndFragment = true;
+
+                if (uri.Segments.Last().StartsWith("$metadata", StringComparison.InvariantCultureIgnoreCase))
```
Comment (habbes) [inline src/Microsoft.OData.CodeGen/Common/MetadataReader.cs:143]: Why using `StartsWith` instead of checking that the strings match exactly? If you use `StartsWith`, that means a segment like `$metadataFoo` will be considered equivalent to `$metadata`. Is that the intent?

File: src/Microsoft.OData.CodeGen/Common/MetadataReader.cs [hunk 5]
```diff
@@ -79,32 +92,76 @@ public static Version GetMetadataVersion(ServiceConfiguration serviceConfigurati
...
+                    var absolutePathUri = new UriBuilder(uri.Scheme, uri.Host, uri.Port, uri.AbsolutePath.TrimEnd('/') + "/").Uri;
+                    uriBuilder = new UriBuilder(new Uri(absolutePathUri, "$metadata"));
+                }
+
+                if (preserveQueryAndFragment)
+                {
+                    uriBuilder.Query = uri.Query.TrimStart('?');
+                    uriBuilder.Fragment = uri.Fragment.TrimStart('#');
+                }
+                uriBuilder.UserName = uri.UserInfo;
```
Comment (habbes) [inline src/Microsoft.OData.CodeGen/Common/MetadataReader.cs:160]: nit: empty line after `}`.

```suggestion

                uriBuilder.UserName = uri.UserInfo;
```

File: test/ODataConnectedService.Tests/ViewModels/ConfigOdataEndPointViewModelTests.cs [hunk 6]
```diff
@@ -54,12 +54,19 @@ public void OnPageLeavingConfigODataEndpointPageTest()
...
+            Assert.AreEqual(configOdataEndPointViewModel.UserSettings.Endpoint, "http://user:password@mysite/ODataService/$metadata?$schemaversion=2.0#fragment");
+
             //Provide a url without $metadata
-            configOdataEndPointViewModel.UserSettings.Endpoint = "http://mysite/ODataService";
+            configOdataEndPointViewModel.UserSettings.Endpoint = "http://user:password@mysite/ODataService?$schemaversion=2.0#fragment";
             pageNavigationResultTask = configOdataEndPointViewModel.OnPageLeavingAsync(null);
 
-            //Check if $metadata is apended if the url does not have it added at the end
-            Assert.AreEqual(configOdataEndPointViewModel.UserSettings.Endpoint, "http://mysite/ODataService/$metadata");
+            //Check if $metadata is apended as the last segment if it was not the last segment of the url
```
Comment (habbes) [inline test/ODataConnectedService.Tests/ViewModels/ConfigOdataEndPointViewModelTests.cs:68]:
```suggestion
            //Check if $metadata is appended as the last segment if it was not the last segment of the url
```

File: test/ODataConnectedService.Tests/ViewModels/ConfigOdataEndPointViewModelTests.cs [hunk 7]
```diff
@@ -54,12 +54,19 @@ public void OnPageLeavingConfigODataEndpointPageTest()
...
+            //Provide a url with "$metadata" as the last segment in the url
+            configOdataEndPointViewModel.UserSettings.Endpoint = "http://user:password@mysite/ODataService/$metadata?$schemaversion=2.0#fragment";
+            _ = configOdataEndPointViewModel.OnPageLeavingAsync(null);
+
+            //Check if the url is detected as valid and is unmodified
+            Assert.AreEqual(configOdataEndPointViewModel.UserSettings.Endpoint, "http://user:password@mysite/ODataService/$metadata?$schemaversion=2.0#fragment");
+
             //Provide a url without $metadata
-            configOdataEndPointViewModel.UserSettings.Endpoint = "http://mysite/ODataService";
+            configOdataEndPointViewModel.UserSettings.Endpoint = "http://user:password@mysite/ODataService?$schemaversion=2.0#fragment";
```
Comment (habbes) [inline test/ODataConnectedService.Tests/ViewModels/ConfigOdataEndPointViewModelTests.cs:65]: I think it would have been better to add new test cases rather than modifying the existing ones, that way we can be confident that current scenarios are still supported.

File: src/Microsoft.OData.CodeGen/Common/MetadataReader.cs (hunk 3 above)
Comment (Nthemba) [reply src/Microsoft.OData.CodeGen/Common/MetadataReader.cs:58]: After a Rebase, changes made in the master branch conflicted with changes in the PR, ServiceConfiguration is used instead of UserSettings

File: src/Microsoft.OData.CodeGen/Common/MetadataReader.cs (hunk 4 above)
Comment (Nthemba) [reply src/Microsoft.OData.CodeGen/Common/MetadataReader.cs:143]: The outcome is still the same i.e. a 404 ... Whether it's a 'StartsWith' or 'Equals'. the code will throw a 404 not found because the URL is not correct. Unless code is implemented to fix segments like '$metadataFoo".

File: azure-pipelines.yml [hunk 8]
```diff
@@ -101,10 +101,10 @@ stages:
...
+      displayName: 'Publish Artifact - VSIX2022Plus'
```
Comment (KenitoInc) [inline azure-pipelines.yml:104]: You should have rebased your branch to make this PR cleaner.
No problem though

File: src/Microsoft.OData.CodeGen/Common/MetadataReader.cs [hunk 9]
```diff
@@ -24,36 +26,47 @@ public static class MetadataReader
...
 
-            
+                uri = CleanMetadataUri(uri);
+
+                serviceConfiguration.Endpoint = uri.AbsoluteUri;
+            }
 
             Stream metadataStream;
-            Uri metadataUri = new Uri(serviceConfiguration.Endpoint);
+            var metadataUri = new Uri(serviceConfiguration.Endpoint);
```
Comment (KenitoInc) [inline src/Microsoft.OData.CodeGen/Common/MetadataReader.cs:51]: why replace `Uri` with `var`?

File: src/Microsoft.OData.CodeGen/Common/MetadataReader.cs [hunk 10]
```diff
@@ -24,36 +26,47 @@ public static class MetadataReader
...
 
             if (!metadataUri.IsFile)
             {
                 var webRequest = (HttpWebRequest)WebRequest.Create(metadataUri);
+                if (serviceConfiguration.CustomHttpHeaders != null)
```
Comment (KenitoInc) [inline src/Microsoft.OData.CodeGen/Common/MetadataReader.cs:56]:
```suggestion

                if (serviceConfiguration.CustomHttpHeaders != null)
```

File: src/Microsoft.OData.CodeGen/Common/MetadataReader.cs [hunk 11]
```diff
@@ -79,32 +92,77 @@ public static Version GetMetadataVersion(ServiceConfiguration serviceConfigurati
...
             }
 
+            var workFile = Path.GetTempFileName();
+
             try
             {
                 using (XmlReader reader = XmlReader.Create(metadataStream))
                 {
-                    while (reader.NodeType != XmlNodeType.Element)
+                    using (var writer = XmlWriter.Create(workFile))
```
Comment (KenitoInc) [inline src/Microsoft.OData.CodeGen/Common/MetadataReader.cs:101]: I don't understand this change.
Why are we calling `writer.WriteNode(reader, false);`?

File: test/ODataConnectedService.Tests/ViewModels/ConfigOdataEndPointViewModelTests.cs [hunk 12]
```diff
@@ -55,11 +55,32 @@ public void OnPageLeavingConfigODataEndpointPageTest()
             Assert.IsTrue(pageNavigationResult.ShowMessageBoxOnFailure);
 
             //Provide a url without $metadata
-            configOdataEndPointViewModel.UserSettings.Endpoint = "http://mysite/ODataService";
+            configOdataEndPointViewModel.UserSettings.Endpoint = "http://mysite/ODataService/$metadata";
```
Comment (habbes) [inline test/ODataConnectedService.Tests/ViewModels/ConfigOdataEndPointViewModelTests.cs:58]: This change contradicts the comment above it. From the above comment, I assume that this test case verified that you can pass a url without $metadata. Maybe you should create a new test case to verify that it works with $metadata, instead of modifying the existing one (unless we already have a unit test verifying that).

File: test/ODataConnectedService.Tests/ViewModels/ConfigOdataEndPointViewModelTests.cs [hunk 13]
```diff
@@ -55,11 +55,32 @@ public void OnPageLeavingConfigODataEndpointPageTest()
...
+            Assert.AreEqual("http://user:password@mysite/ODataService/$metadata?$schemaversion=2.0#fragment", configOdataEndPointViewModel.ServiceConfiguration.Endpoint);
+
+            //Provide a url with query and fragment segments without $metadata
+            configOdataEndPointViewModel.UserSettings.Endpoint = "http://user:password@mysite/ODataService?$schemaversion=2.0#fragment";
+            pageNavigationResultTask = configOdataEndPointViewModel.OnPageLeavingAsync(null);
+
+            //Check if $metadata is appended as the last segment if it was not the last segment of the url
+            Assert.AreEqual("http://user:password@mysite/ODataService/$metadata?$schemaversion=2.0#fragment", configOdataEndPointViewModel.ServiceConfiguration.Endpoint);
+
+            //Provide a url with a fragment segment without $metadata
```
Comment (habbes) [inline test/ODataConnectedService.Tests/ViewModels/ConfigOdataEndPointViewModelTests.cs:78]: I think you could also add a test case that provides a url with $metadata and #fragment without a query string.

File: test/ODataConnectedService.Tests/ViewModels/ConfigOdataEndPointViewModelTests.cs [hunk 14]
```diff
@@ -55,11 +55,32 @@ public void OnPageLeavingConfigODataEndpointPageTest()
...
-            configOdataEndPointViewModel.UserSettings.Endpoint = "http://mysite/ODataService";
+            configOdataEndPointViewModel.UserSettings.Endpoint = "http://mysite/ODataService/$metadata";
             pageNavigationResultTask = configOdataEndPointViewModel.OnPageLeavingAsync(null);
 
-            //Check if $metadata is apended if the url does not have it added at the end
-            Assert.AreEqual(configOdataEndPointViewModel.UserSettings.Endpoint, "http://mysite/ODataService/$metadata");
+            //Check if $metadata is appended as the last segment if it was not the last segment of the url
+            Assert.AreEqual("http://mysite/ODataService/$metadata", configOdataEndPointViewModel.ServiceConfiguration.Endpoint);
+
+            //Provide a url with "$metadata" as the last segment in the url
```
Comment (habbes) [inline test/ODataConnectedService.Tests/ViewModels/ConfigOdataEndPointViewModelTests.cs:64]:
```suggestion
            //Provide a url  with a query string and "$metadata" as the last segment in the url path
```

**Comment by Nthemba** [comment]
Find updates to this PR on the link below
https://github.com/OData/ODataConnectedService/pull/311/
